prov/efa/src/rxr/rxr_pkt_type_base.c

```diff
@@ -338,6 +338,109 @@ int rxr_pkt_queued_copy_data_to_hmem(struct rxr_ep *ep,
 	return rxr_ep_flush_queued_blocking_copy_to_hmem(ep);
 }
 
+/* @breif copy data in pkt_entry to CUDA memory
+ *
+ * CUDA is different from other types of device memory in two ways:
+ *
+ * First, we might not be able to use ofi_copy_iov_hmem_iov(). This is because
+ * that function may call cudaMemcpy(), when gdrcopy is not available, and
+ * Nvidia Collective Communications Library (NCCL) forbids its plugins (hence
+ * libfabric) to call cudaMemcpy. (Doing so can result in a deadlock). so
+ * ofi_copy_iov_hmem_iov() can only be used when gdrcopy is available.
+ *
+ * Second, there is another method to copy data: local RDMA read based copy.
+ * This method is slower than gdrcopy in some cases, but it does not consume
+ * CPU resources so we want to use it when possible.
+ *
+ * This function picks the proper method to perform the copy to CUDA memory.
+ * The decide is made based on multiple factors:
+ *
+ * 1) whether gdrcopy is available.
+ * 2) whether data in the packet is the last piece
+ * 3) number of inflight RX entry that is using gdrcopy.
+ *
+ * @param[in]		ep		endpoint
+ * @param[in]		pkt_entry	the packet entry that contains data, which
+ *                                      x_entry pointing to the correct rx_entry.
+ * @param[in]		data		the pointer pointing to the beginning of data
+ * @param[in]		data_size	the length of data
+ * @param[in]		data_offset	the offset of the data in the packet in respect
+ *					of the receiving buffer.
+ * @return		On success, return 0
+ * 			On failure, return libfabric error code
+ */
+static inline
+int rxr_pkt_copy_data_to_cuda(struct rxr_ep *ep,
+			      struct rxr_pkt_entry *pkt_entry,
+			      char *data,
+			      size_t data_size,
+			      size_t data_offset)
+{
+	static const int max_gdrcopy_rx_entry_num = 4;
+	struct rxr_rx_entry *rx_entry;
+	struct efa_mr *desc;
+	struct efa_ep *efa_ep;
+	int use_p2p, err;
+
+	rx_entry = pkt_entry->x_entry;
+	desc = rx_entry->desc[0];
+	assert(efa_mr_is_cuda(desc));
+
+	if (cuda_is_gdrcopy_enabled() && rx_entry->cuda_copy_method != RXR_CUDA_COPY_LOCALREAD) {
+		assert(rx_entry->bytes_copied + data_size <= rx_entry->total_len);
+
+		/* If this packet is the last uncopied piece (or the only piece), copy it right away
+		 * to achieve best latency.
+		 */
+		if (rx_entry->bytes_copied + data_size == rx_entry->total_len) {
+			ofi_copy_to_hmem_iov(desc->peer.iface, desc->peer.device.reserved,
+					     rx_entry->iov, rx_entry->iov_count,
+					     data_offset + ep->msg_prefix_size,
+					     data, data_size);
+			rxr_pkt_handle_data_copied(ep, pkt_entry, data_size);
```

Comment:
I think it's a little unintuitive that ```ep->gdrcopy_rx_entry_num``` doesn't change and the copy type is left as UNSPEC, but it looks logically correct at least.

Reply:
This is the last piece of data. With this copy the receive is completed. so if we increase `gdrcopy_rx_entry_num` here, it will be immediately reduced in `rxr_pkt_handle_data_copied`.

Reply:
Yeah it should work logically, I just found it slightly unintuitive